src/Course/Functor.hs

```diff
@@ -10,119 +11,130 @@ import Course.List
-    (<$>) = error "todo: Course.Functor (<$>)#instance ExactlyOne"
-
--- | Maps a function on the List functor.
---
--- >>> (+1) <$> Nil
--- []
---
--- >>> (+1) <$> (1 :. 2 :. 3 :. Nil)
--- [2,3,4]
+    (<$>) f (ExactlyOne x) = ExactlyOne $ f x
+
+{- | Maps a function on the List functor.
+
+>>> (+1) <$> Nil
+[]
+
+>>> (+1) <$> (1 :. 2 :. 3 :. Nil)
+[2,3,4]
+-}
 instance Functor List where
     (<$>) :: (a -> b) -> List a -> List b
-    (<$>) = error "todo: Course.Functor (<$>)#instance List"
-
--- | Maps a function on the Optional functor.
---
--- >>> (+1) <$> Empty
--- Empty
---
--- >>> (+1) <$> Full 2
--- Full 3
+    (<$>) f = foldRight (\y acc -> f y :. acc) Nil
+
+{- | Maps a function on the Optional functor.
+
+>>> (+1) <$> Empty
+Empty
+
+>>> (+1) <$> Full 2
+Full 3
+-}
 instance Functor Optional where
     (<$>) :: (a -> b) -> Optional a -> Optional b
-    (<$>) = error "todo: Course.Functor (<$>)#instance Optional"
+    (<$>) _ Empty = Empty
+    (<$>) f (Full x) = Full $ f x
+
+{- | Maps a function on the reader ((->) t) functor.
 
--- | Maps a function on the reader ((->) t) functor.
---
--- >>> ((+1) <$> (*2)) 8
--- 17
+>>> ((+1) <$> (*2)) 8
+17
+-}
 instance Functor ((->) t) where
-    (<$>) :: (a -> b) -> ((->) t a) -> ((->) t b)
-    (<$>) = error "todo: Course.Functor (<$>)#((->) t)"
-
--- | Anonymous map. Maps a constant value on a functor.
---
--- >>> 7 <$ (1 :. 2 :. 3 :. Nil)
--- [7,7,7]
---
--- prop> \x a b c -> x <$ (a :. b :. c :. Nil) == (x :. x :. x :. Nil)
---
--- prop> \x q -> x <$ Full q == Full x
+    (<$>) :: (a -> b) -> (->) t a -> (->) t b
+    (<$>) f g = f . g
+
+{- | Anonymous map. Maps a constant value on a functor.
+
+>>> 7 <$ (1 :. 2 :. 3 :. Nil)
+[7,7,7]
+
+prop> \x a b c -> x <$ (a :. b :. c :. Nil) == (x :. x :. x :. Nil)
+
+prop> \x q -> x <$ Full q == Full x
+-}
 (<$) :: Functor k => a -> k b -> k a
-(<$) = error "todo: Course.Functor#(<$)"
-
--- | Apply a value to a functor-of-functions.
---
--- __NOTE__: The second argument is a bare @a@, not a @k a@. You need
--- a more powerful typeclass, 'Applicative', if you want both the
--- functions and the argmuents to be "inside" the Functor:
---
--- @
--- (<*>) :: Applicative k => k (a -> b) -> k a -> k b
--- @
---
--- We will talk about 'Applicative' soon.
---
--- >>> (*2) :. (+1) :. const 99 :. Nil ?? 8
--- [16,9,99]
---
--- >>> Empty ?? 2
--- Empty
+(<$) x = (<$>) (const x)
+
+{- | Apply a value to a functor-of-functions.
+
+ __NOTE__: The second argument is a bare @a@, not a @k a@. You need
+ a more powerful typeclass, 'Applicative', if you want both the
+ functions and the argmuents to be "inside" the Functor:
+
+ @
+ (<*>) :: Applicative k => k (a -> b) -> k a -> k b
+ @
+
+ We will talk about 'Applicative' soon.
+
+>>> (*2) :. (+1) :. const 99 :. Nil ?? 8
+[16,9,99]
+
+>>> Empty ?? 2
+Empty
+-}
 (??) :: Functor k => k (a -> b) -> a -> k b
-(??) ff a = error "todo: Course.Functor#(??)"
+(??) f a = (<$>) (\x -> x a) f
```

Comment:
```suggestion
(??) f a = (<$>) ($ a) f
```

> Avoid lambdas at all cost!

Reply:
Wow I never even though of using `$` that way...